tests/test_flask.py

```diff
@@ -79,6 +80,52 @@ def teardown():
             yield client
             teardown()
 
+def test_failed_login(client):
+    #Testing the failed login page
+    with pytest.raises(InvalidCharactersException):
+        returned_value = client.post('/login', data=dict(
+            username="testTrainee#%#^",
+            password="password#^#$#"
+        ), follow_redirects=True)
+        assert returned_value.status_code == 200
```

Comment:
Split this into two different blocks. 

One `with pytest.raises(InvalidCharactersException): ...` with the username with invalid characters and the password as normal, and another `with pytest.raises(InvalidCharactersException): ...` block with the username as normal and the password with invalid characters. 

Reason is because right now the if statement that checks the username would always raise the exception since it's the first check, but we need to check if the password if statement works as well.